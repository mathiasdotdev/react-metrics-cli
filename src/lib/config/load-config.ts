import fs from 'node:fs';
import * as os from 'os';
import * as path from 'path';
import { getDefaultConfig, ReactMetricsConfig } from './get-default-config';

const CONFIG_DIR = path.join(os.homedir(), '.react-metrics-cli');
const CONFIG_FILE = path.join(CONFIG_DIR, 'config.json');

/**
 * Vérifie si le fichier de configuration existe
 */
export const configExists = (): boolean => {
  return fs.existsSync(CONFIG_FILE);
};

/**
 * Charge la configuration depuis le fichier
 */
export const loadConfig = (): ReactMetricsConfig => {
  if (!configExists()) {
    return getDefaultConfig();
  }

  try {
    const configData = fs.readFileSync(CONFIG_FILE, 'utf-8');
    const config = JSON.parse(configData) as ReactMetricsConfig;

    // Fusionner avec les valeurs par défaut pour s'assurer que tous les champs sont présents
    return { ...getDefaultConfig(), ...config };
  } catch (error) {
    console.error(
      'Erreur lors du chargement de la configuration, utilisation des valeurs par défaut:',
      error,
    );
    return getDefaultConfig();
  }
};

/**
 * Sauvegarde la configuration dans le fichier
 */
export const saveConfig = (config: ReactMetricsConfig): void => {
  // Créer le répertoire si nécessaire
  if (!fs.existsSync(CONFIG_DIR)) {
    fs.mkdirSync(CONFIG_DIR, { recursive: true });
  }

  // Écrire le fichier de configuration
  const configData = JSON.stringify(config, null, 2);
  fs.writeFileSync(CONFIG_FILE, configData, 'utf-8');
};

/**
 * Initialise le fichier de configuration avec les valeurs par défaut
 */
export const initConfig = (): void => {
  const defaultConfig = getDefaultConfig();
  saveConfig(defaultConfig);
};

/**
 * Obtient le chemin du fichier de configuration
 */
export const getConfigPath = (): string => {
  return CONFIG_FILE;
};

/**
 * Supprime le fichier de configuration
 */
export const resetConfig = (): void => {
  if (configExists()) {
    fs.unlinkSync(CONFIG_FILE);
  }
};
