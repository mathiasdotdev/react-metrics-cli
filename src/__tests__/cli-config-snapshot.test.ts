import { afterEach, beforeEach, describe, expect, test } from 'bun:test';
import { existsSync, unlinkSync } from 'node:fs';
import { getConfigPath } from '$config/load-config';
import { runAndNormalizeCliOutput } from './helpers/snapshotTestHelper';

describe('CLI Config Command - Snapshot Tests', () => {
  const configPath = getConfigPath();

  // Nettoyer le fichier de config avant et après chaque test
  beforeEach(() => {
    if (existsSync(configPath)) {
      unlinkSync(configPath);
    }
  });

  afterEach(() => {
    if (existsSync(configPath)) {
      unlinkSync(configPath);
    }
  });

  test(
    'config --init devrait initialiser le fichier de configuration',
    async () => {
      const normalizedOutput = await runAndNormalizeCliOutput(
        ['config', '--init'],
        { timeout: 30000 },
        { normalizePaths: true, removeAnsiCodes: true },
      );

      expect(normalizedOutput).toMatchSnapshot();
    },
    { timeout: 30000 },
  );

  test(
    'config --info devrait afficher les informations avec fichier existant',
    async () => {
      // Créer d'abord le fichier de config
      await runAndNormalizeCliOutput(['config', '--init'], { timeout: 30000 });

      // Afficher les infos
      const normalizedOutput = await runAndNormalizeCliOutput(
        ['config', '--info'],
        { timeout: 30000 },
        { normalizePaths: true, removeAnsiCodes: true },
      );

      expect(normalizedOutput).toMatchSnapshot();
    },
    { timeout: 30000 },
  );

  test(
    "config --info devrait afficher un message si le fichier n'existe pas",
    async () => {
      const normalizedOutput = await runAndNormalizeCliOutput(
        ['config', '--info'],
        { timeout: 30000 },
        { normalizePaths: true, removeAnsiCodes: true },
      );

      expect(normalizedOutput).toMatchSnapshot();
    },
    { timeout: 30000 },
  );

  test(
    "config (sans option) devrait afficher l'aide de configuration",
    async () => {
      const normalizedOutput = await runAndNormalizeCliOutput(
        ['config'],
        { timeout: 30000 },
        { normalizePaths: true, removeAnsiCodes: true },
      );

      expect(normalizedOutput).toMatchSnapshot();
    },
    { timeout: 30000 },
  );

  test(
    "config --help devrait afficher l'aide de configuration",
    async () => {
      const normalizedOutput = await runAndNormalizeCliOutput(
        ['config', '--help'],
        { timeout: 30000 },
        { normalizePaths: true, removeAnsiCodes: true },
      );

      expect(normalizedOutput).toMatchSnapshot();
    },
    { timeout: 30000 },
  );
});
