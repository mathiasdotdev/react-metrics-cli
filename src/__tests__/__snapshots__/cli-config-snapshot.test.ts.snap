// Bun Snapshot v1, https://bun.sh/docs/test/snapshots

exports[`CLI Config Command - Snapshot Tests config --init devrait initialiser le fichier de configuration 1`] = `
"╭────────────────────────────────────╮
│         React Metrics CLI          │
│       Analyseur de code mort       │
╰────────────────────────────────────╯


✅ Fichier de configuration initialisé: C:/Users/mathl/.react-metrics-cli/config.json
# Modifiez le fichier directement dans votre éditeur"
`;

exports[`CLI Config Command - Snapshot Tests config --info devrait afficher les informations avec fichier existant 1`] = `
"╭────────────────────────────────────╮
│         React Metrics CLI          │
│       Analyseur de code mort       │
╰────────────────────────────────────╯


💡 Informations de configuration React-Metrics

📁 Fichier: C:/Users/mathl/.react-metrics-cli/config.json

⚙️ Configuration actuelle:
- Extensions de fichiers: .js, .jsx, .ts, .tsx
- Dossiers ignorés: node_modules, .git, dist, build, coverage
- Ignorer annotations: Oui
📊 Rapports:
- Terminal: Activé
- HTML: Désactivé
- JSON: Désactivé
🔍 Analyses:
- Constantes: Activée
- Fonctions: Activée
- Classes: Activée
- Props: Activée
- Consoles: Activée
- Definitions: Activée
- Imports: Activée
- Exports: Activée
- Dépendances: Désactivée"
`;

exports[`CLI Config Command - Snapshot Tests config --info devrait afficher un message si le fichier n'existe pas 1`] = `
"╭────────────────────────────────────╮
│         React Metrics CLI          │
│       Analyseur de code mort       │
╰────────────────────────────────────╯


💡 Informations de configuration React-Metrics

💡 Utilisez 'react-metrics config --init' pour créer un fichier de configuration
⚠️ Aucun fichier de configuration trouvé"
`;

exports[`CLI Config Command - Snapshot Tests config (sans option) devrait afficher l'aide de configuration 1`] = `
"╭────────────────────────────────────╮
│         React Metrics CLI          │
│       Analyseur de code mort       │
╰────────────────────────────────────╯


⚙️ Configuration React-Metrics

💡 Utilisez --init pour créer un fichier de configuration

📁 Emplacement: C:/Users/mathl/.react-metrics-cli/config.json

⚙️ Options disponibles:
- --info              Afficher la configuration actuelle
- --init              Créer/réinitialiser le fichier de configuration
- --open              Ouvrir le fichier de configuration dans VS Code


⚙️ # Modification manuelle:
- Ouvrez le fichier dans votre éditeur pour le modifier
- Le fichier sera automatiquement utilisé par react-metrics


✍️ Exemples:
- react-metrics config --info
- react-metrics config --init
- react-metrics config --open
- code C:/Users/mathl/.react-metrics-cli/config.json
⚠️ Aucun fichier de configuration trouvé"
`;

exports[`CLI Config Command - Snapshot Tests config --help devrait afficher l'aide de configuration 1`] = `
"╭────────────────────────────────────╮
│         React Metrics CLI          │
│       Analyseur de code mort       │
╰────────────────────────────────────╯

# Usage:
  react-metrics config [options]

# Description:
  Gérer la configuration React-Metrics

# Options:
  --init               Initialiser le fichier de configuration avec les valeurs par défaut
  -i, --info           Afficher les informations de configuration
  -o, --open           Ouvrir le fichier de configuration dans VS Code
  -h, --help           display help for command"
`;
